mobilecoind/src/sync.rs

```diff
@@ -603,6 +615,193 @@ mod test {
         assert!(!utxos.contains(&first_utxo));
     }
 
+    // TODO: make this a bench instead of a unit test.
+    #[test_with_logger]
+    fn test_sync_monitor_with_random_recipients(logger: Logger) {
+        let mut rng: StdRng = SeedableRng::from_seed([98u8; 32]);
+
+        let account_keys: Vec<_> = (0..5).map(|_i| AccountKey::random(&mut rng)).collect();
+
+        let data = MonitorData::new(
+            account_keys[0].clone(),
+            DEFAULT_SUBADDRESS_INDEX, // first subaddress
+            5,                        // number of subaddresses
+            0,                        // first block
+            "",                       // name
+        )
+        .unwrap();
+
+        let monitor_id = MonitorId::from(&data);
+
+        let recipients: Vec<PublicAddress> = account_keys
+            .iter()
+            .map(AccountKey::default_subaddress)
+            .collect();
+
+        // Generate a test database with a number blocks that does not divide evenly by
+        // MAX_BLOCKS_PROCESSING_CHUNK_SIZE.
+        let num_blocks = (MAX_BLOCKS_PROCESSING_CHUNK_SIZE * 2) + 1;
+        let (mut ledger_db, mobilecoind_db) = get_test_databases(
+            BlockVersion::MAX,
+            0,
+            &recipients,
+            num_blocks,
+            logger.clone(),
+            &mut rng,
+        );
+
+        // Our recipient (controlled by the monitor id) is the first account
+        // (account_keys[0]). Each block generated by test_utils has a TxOut per
+        // recipient, so building on that knowledge the following code gets us
+        // the TxOuts relevant to our particular recipient.
+        let account0_tx_outs: Vec<TxOut> = (0..num_blocks)
+            .map(|idx| {
+                let block_contents = ledger_db.get_block_contents(idx as u64).unwrap();
+                block_contents.outputs[0].clone()
+            })
+            .collect();
+
+        // Before doing anything, we should not have any utxos for our test monitor.
+        let utxos = mobilecoind_db
+            .get_utxos_for_subaddress(&monitor_id, DEFAULT_SUBADDRESS_INDEX)
+            .unwrap();
+        assert_eq!(utxos.len(), 0);
+
+        // Add monitor, should still have 0 outputs.
+        assert_eq!(mobilecoind_db.add_monitor(&data).unwrap(), monitor_id);
+
+        // Haven't synced yet, so still no outputs expected.
+        let utxos = mobilecoind_db
+            .get_utxos_for_subaddress(&monitor_id, DEFAULT_SUBADDRESS_INDEX)
+            .unwrap();
+        assert_eq!(utxos.len(), 0);
+
+        let monitor_data = mobilecoind_db.get_monitor_data(&monitor_id).unwrap();
+        assert_eq!(monitor_data.next_block, 0);
+
+        // Process the first MAX_BLOCKS_PROCESSING_CHUNK_SIZE blocks.
+        let result = sync_monitor(&ledger_db, &mobilecoind_db, &monitor_id, &logger).unwrap();
+        assert_eq!(result, SyncMonitorOk::MoreBlocksPotentiallyAvailable);
```

Comment:
This logic is reused multiple times. It might make sense to factor it out into a helper method in a future PR.